File: apps/mobile/modules/rishi-pdf-extractor/ios/PdfExtractor.swift
import Foundation
import PDFKit

public struct WordRect {
  public let idx: Int
  public let text: String
  public let x: Double
  public let y: Double
  public let w: Double
  public let h: Double
}

public struct PageData {
  public let pageNumber: Int
  public let widthPts: Double
  public let heightPts: Double
  public let paragraphs: [ParagraphOut]
  public let words: [WordRect]   // empty for now — filled in Task 6
}

public enum PdfExtractorError: Error {
  case fileNotFound(String)
  case pageNotFound(Int)
}

public enum PdfExtractor {
  public static func getPageCount(path: String) throws -> Int {
    guard let doc = PDFDocument(url: URL(fileURLWithPath: path)) else {
      throw PdfExtractorError.fileNotFound(path)
    }
    return doc.pageCount
  }

  public static func extractPage(path: String, pageNumber: Int) throws -> PageData {
    guard let doc = PDFDocument(url: URL(fileURLWithPath: path)) else {
      throw PdfExtractorError.fileNotFound(path)
    }
    // Page numbers are 1-based in our JS API; PDFKit is 0-based.
    let idx = pageNumber - 1
    guard idx >= 0 && idx < doc.pageCount, let page = doc.page(at: idx) else {
      throw PdfExtractorError.pageNotFound(pageNumber)
    }

    let bounds = page.bounds(for: .mediaBox)
    let charCount = page.numberOfCharacters

    // PDFPage does not expose a public character(at:) returning unichar.
    // We use the page's full string and characterBounds(at:) together.
    let pageString = page.string ?? ""
    let scalars = Array(pageString.unicodeScalars)

    var lines: [Line] = []
    var currentLineText = ""
    var currentLineMinY: Double = .infinity
    var currentLineMaxY: Double = -.infinity
    var prevLineKey: Int = Int.min

    // Use min(charCount, scalars.count) to guard against any length mismatch.
    let count = min(charCount, scalars.count)

    for i in 0..<count {
      let scalar = scalars[i]
      // Skip carriage returns / newline control characters used by PDFKit
      // as line-break markers; we derive line breaks from Y-position instead.
      if scalar.value == 0x000A || scalar.value == 0x000D { continue }

      let rect = page.characterBounds(at: i)
      // Guard against zero-size rects (invisible/control characters).
      guard rect.width > 0 || rect.height > 0 else { continue }

      // Group characters into visual lines by their Y-row bucket.
      // Dividing midY by max(height,1) gives a stable integer key per row.
      let charLineKey = Int((rect.midY / max(rect.height, 1.0)).rounded())

      if charLineKey != prevLineKey && prevLineKey != Int.min {
        if !currentLineText.isEmpty {
          lines.append(Line(
            text: currentLineText,
            originY: currentLineMinY,
            height: max(currentLineMaxY - currentLineMinY, 1)))
        }
        currentLineText = ""
        currentLineMinY = .infinity
        currentLineMaxY = -.infinity
      }

      currentLineText.unicodeScalars.append(scalar)
      currentLineMinY = min(currentLineMinY, Double(rect.minY))
      currentLineMaxY = max(currentLineMaxY, Double(rect.maxY))
      prevLineKey = charLineKey
    }
    // Flush final line.
    if !currentLineText.isEmpty {
      lines.append(Line(
        text: currentLineText,
        originY: currentLineMinY,
        height: max(currentLineMaxY - currentLineMinY, 12)))
    }

    let paragraphs = ParagraphSegmenter.segment(lines, pageNumber: pageNumber)

    return PageData(
      pageNumber: pageNumber,
      widthPts: Double(bounds.width),
      heightPts: Double(bounds.height),
      paragraphs: paragraphs,
      words: []
    )
  }
}

File: apps/mobile/modules/rishi-pdf-extractor/android/src/main/java/expo/modules/rishipdfextractor/PdfExtractor.kt
package expo.modules.rishipdfextractor

import android.content.Context
import android.os.ParcelFileDescriptor
import com.shockwave.pdfium.PdfiumCore
import com.shockwave.pdfium.PdfDocument
import com.shockwave.pdfium.util.Size
import java.io.File
import kotlin.math.max
import kotlin.math.min

data class WordRectOut(
  val idx: Int,
  val text: String,
  val x: Double, val y: Double, val w: Double, val h: Double,
)

data class PageDataOut(
  val pageNumber: Int,
  val widthPts: Double,
  val heightPts: Double,
  val paragraphs: List<ParagraphOut>,
  val words: List<WordRectOut>,
)

class PdfExtractor(private val ctx: Context) {

  private fun openDoc(path: String): Pair<PdfiumCore, PdfDocument> {
    val core = PdfiumCore(ctx)
    val fd = ParcelFileDescriptor.open(File(path), ParcelFileDescriptor.MODE_READ_ONLY)
    val doc = core.newDocument(fd)
    return core to doc
  }

  fun getPageCount(path: String): Int {
    val (core, doc) = openDoc(path)
    try {
      return core.getPageCount(doc)
    } finally {
      core.closeDocument(doc)
    }
  }

  fun extractPage(path: String, pageNumber: Int): PageDataOut {
    val (core, doc) = openDoc(path)
    try {
      val idx = pageNumber - 1
      core.openPage(doc, idx)
      val size: Size = core.getPageSize(doc, idx)
      val widthPts = size.width.toDouble()
      val heightPts = size.height.toDouble()

      val textPagePtr = core.openTextPage(doc, idx)
      val charCount = core.textPageCountChars(textPagePtr)

      val lines = mutableListOf<Line>()
      var currentText = StringBuilder()
      var currentMinY = Double.POSITIVE_INFINITY
      var currentMaxY = Double.NEGATIVE_INFINITY
      var prevLineKey = Int.MIN_VALUE

      for (i in 0 until charCount) {
        val unicode = core.textPageGetUnicode(textPagePtr, i)
        // Skip CR/LF — line breaks come from Y-grouping.
        if (unicode == 0x000A || unicode == 0x000D) continue

        val rect = core.textPageGetCharBox(textPagePtr, i)
        // Pdfium reports rects in PDF coords; .top may be greater than .bottom.
        val minY = min(rect.top.toDouble(), rect.bottom.toDouble())
        val maxY = max(rect.top.toDouble(), rect.bottom.toDouble())
        val height = maxY - minY
        // Guard against zero-size rects.
        if (height <= 0) continue

        val midY = (minY + maxY) / 2
        val key = (midY / max(height, 1.0)).toInt()

        if (key != prevLineKey && prevLineKey != Int.MIN_VALUE) {
          if (currentText.isNotEmpty()) {
            lines += Line(
              text = currentText.toString(),
              originY = currentMinY,
              height = max(currentMaxY - currentMinY, 1.0),
            )
          }
          currentText = StringBuilder()
          currentMinY = Double.POSITIVE_INFINITY
          currentMaxY = Double.NEGATIVE_INFINITY
        }
        currentText.append(unicode.toChar())
        currentMinY = min(currentMinY, minY)
        currentMaxY = max(currentMaxY, maxY)
        prevLineKey = key
      }
      if (currentText.isNotEmpty()) {
        lines += Line(
          text = currentText.toString(),
          originY = currentMinY,
          height = max(currentMaxY - currentMinY, 12.0),
        )
      }
      core.closeTextPage(textPagePtr)

      val paragraphs = ParagraphSegmenter.segment(lines, pageNumber)
      return PageDataOut(
        pageNumber = pageNumber,
        widthPts = widthPts,
        heightPts = heightPts,
        paragraphs = paragraphs,
        words = emptyList(),  // filled in Task 7
      )
    } finally {
      core.closeDocument(doc)
    }
  }
}
